### PR TITLE
ci: refactor Earthly

### PR DESCRIPTION
**Pull request by nrdxp**
# Description

Included are some refactors that should make our Earthly workflow easier to comprehend.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages.
- [ ] New tests are added if needed and existing tests are updated.
- [ ] Relevant logging and metrics added
- [ ] CI passes. See note on CI.
- [ ] Any changes are noted in the `changelog.md` for affected crate
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG Partner Chains developers to do this
for you.

### CODE REVIEW

**Review by skylar-simoncelli — APPROVED** [review]

**Review by tgunnoe — APPROVED** [review]
Thank you for the organization and especially the comments/documentation around the Earthfile.

File: Earthfile [hunk 1]
```diff
@@ -99,57 +59,74 @@ docker:
...
-  FOR crate IN $CRATES
-    COPY --if-exists $crate/Cargo.toml $crate/Cargo.toml
+test:
+  FROM +build
+  LET WASM_BUILD_STD=0
+  DO github.com/earthly/lib:3.0.3+INSTALL_DIND
+  CACHE --sharing shared --id cargo $CARGO_HOME
+  RUN cargo test --no-run --locked --profile=$PROFILE --features=$FEATURES,runtime-benchmarks
+  WITH DOCKER
+    RUN cargo test --locked --profile=$PROFILE --features=$FEATURES,runtime-benchmarks
```
Comment (LGLO) [inline Earthfile:69]: I know it's not new here, but why do we have to `cargo test --no-run` before running the actual tests?

**Review by LGLO — APPROVED** [review]

File: Earthfile (hunk 1 above)
Comment (nrdxp) [reply Earthfile:69]: Just another layer optmization, to split building the test executables from actually running them.

In general in docer setups (not just earthly), it is typical and good practice to split steps into descreet units to maximize the layer caches effectiveness.

**Comment by tgunnoe** [comment]
can we rebase/merge this?

**Comment by AmbientTea** [comment]
Is this PR still relevant?

**Review by AmbientTea — COMMENTED** [review]
Is this PR still relevant?

**Comment by nrdxp** [comment]
I think it is, yes, however I don't want to cause @skylar-simoncelli any trouble during is CI improvement efforts, so I'll make sure he is done before trying to rebase and merge this, it might require some modification afterwards as well

**Comment by nrdxp** [comment]
cc @skylar-simoncelli, I had some time to go ahead and rebase this after your changes, it ends up being mostly the same after resolving the merge conflicts, however there was an opportunity to simplify the ghcr upload using earthly directly. perhaps we need to trigger manually to ensure it works, although it should.

**Comment by skylar-simoncelli** [comment]
Great job with the rebase and the restructure! I like it

**Comment by nrdxp** [comment]
@skylar-simoncelli, any insight as to why this might be failing?
https://github.com/input-output-hk/partner-chains/actions/runs/11410102634/job/31752614698#step:3:399

**Comment by skylar-simoncelli** [comment]
@nrdxp the pod is in a crash loop:

```
$ kubectl get all -n integration-testing
NAME                                                                  READY   STATUS             RESTARTS      AGE
pod/substrate-node-sha-74d9edc606633d40c7232dbccf3ff4c7becfd53qkzh7   0/1     CrashLoopBackOff   3 (18s ago)   64s

NAME                                                           TYPE       CLUSTER-IP     EXTERNAL-IP   PORT(S)                                         AGE
service/sha-74d9edc606633d40c7232dbccf3ff4c7becfd53c-service   NodePort   172.20.83.66   <none>        30333:30494/TCP,9933:31265/TCP,9615:31806/TCP   63s

NAME                                                                          READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/substrate-node-sha-74d9edc606633d40c7232dbccf3ff4c7becfd53c   0/1     1            0           65s

NAME                                                                                     DESIRED   CURRENT   READY   AGE
replicaset.apps/substrate-node-sha-74d9edc606633d40c7232dbccf3ff4c7becfd53c-6cc657f88d   1         1         0       65s
```

Here is the cause:


```
$ kubectl describe pod substrate-node-sha-74d9edc606633d40c7232dbccf3ff4c7becfd53qkzh7 -n integration-testing
...
Error: failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "./usr/local/bin/$NODE": stat ./usr/local/bin/$NODE: no such file or directory: unknown
...

**Pull request closed without merging**